### PR TITLE
Fix array == null error in AccessibilityBridge

### PR DESCRIPTION
**Pull request by GuaiYiHu**
08-16 06:23:59.641 23012 23012 E flutter : [ERROR:flutter/fml/platform/android/jni_util.cc(182)] java.lang.IllegalArgumentException: array == null
	at android.opengl.Matrix.multiplyMM(Native Method)
	at io.flutter.view.AccessibilityBridge$SemanticsNode.updateRecursively(Unknown Source:28)
	at io.flutter.view.AccessibilityBridge$SemanticsNode.updateRecursively(Unknown Source:240)
	at io.flutter.view.AccessibilityBridge$SemanticsNode.updateRecursively(Unknown Source:240)
	at io.flutter.view.AccessibilityBridge$SemanticsNode.updateRecursively(Unknown Source:240)
	at io.flutter.view.AccessibilityBridge$SemanticsNode.updateRecursively(Unknown Source:240)
	at io.flutter.view.AccessibilityBridge$SemanticsNode.access$4900(Unknown Source:0)
	at io.flutter.view.AccessibilityBridge.updateSemantics(Unknown Source:192)
	at io.flutter.view.AccessibilityBridge$1.updateSemantics(Unknown Source:21)
	at io.flutter.embedding.engine.FlutterJNI.updateSemantics(Unknown Source:9)
	at android.os.MessageQueue.nativePollOnce(Native 
08-16 06:23:59.641 23012 23012 F flutter : [FATAL:flutter/shell/platform/android/platform_view_android_jni_impl.cc(1214)] Check failed: fml::jni::CheckException(env).

Fix array == null error

https://github.com/flutter/flutter/issues/89834
https://github.com/flutter/flutter/issues/31139

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by GuaiYiHu** [comment]
Will any one review this commit?

File: shell/platform/android/io/flutter/view/AccessibilityBridge.java [hunk 1]
```diff
@@ -2642,6 +2642,9 @@ private void updateRecursively(
         if (globalTransform == null) {
           globalTransform = new float[16];
         }
+        if (transform == null) {
+          transform = new float[16];
```
Comment (chinmaygarde) [inline shell/platform/android/io/flutter/view/AccessibilityBridge.java:2646]: Shouldn't this and the thing above be initialized to an identity transformation?

Comment (chinmaygarde) [reply shell/platform/android/io/flutter/view/AccessibilityBridge.java:2646]: cc @Hixie who may have context as the author of the original null check that initialized all matrix elements to zer.

Comment (GuaiYiHu) [reply shell/platform/android/io/flutter/view/AccessibilityBridge.java:2646]: better than crash

Comment (GuaiYiHu) [reply shell/platform/android/io/flutter/view/AccessibilityBridge.java:2646]: > cc @Hixie who may have context as the author of the original null check that initialized all matrix elements to zero.

Hope merge, plz

**Comment by zanderso** [comment]
From PR triage: @stuartmorgan Do you know who would be the right reviewer for this change? Should the concern from @chinmaygarde be addressed?

**Comment by stuartmorgan** [comment]
@chunhtai Are you familiar with this code? It looks like you are one of the people who has been most active in this file.

File: shell/platform/android/io/flutter/view/AccessibilityBridge.java (hunk 1 above)
Comment (chunhtai) [reply shell/platform/android/io/flutter/view/AccessibilityBridge.java:2646]: This should be initialized during the SemanticsNode.updateWith. The only possible case I can think of is the framework sent a semantics node with id child that is not in the update list. This feels like a bug in the framework side. Do you have a reproducible code?

**Comment by GaryQian** [comment]
@GuaiYiHu Can you provide a reproduction of the crash so we can investigate what the correct course of action is? Either mitigate the crash with this PR or fix the upstream root bug.

**Comment by chinmaygarde** [comment]
I think this is a safe fix either way and the same pattern followed elsewhere is being followed here.

**Comment by chunhtai** [comment]
I am ok if we add an assert so that we still have a chance to catch the root cause but avoiding crash in production. @GuaiYiHu  Can you add an assert that it should never be null?

**Comment by GuaiYiHu** [comment]
> I am ok if we add an assert so that we still have a chance to catch the root cause but avoiding crash in production. @GuaiYiHu Can you add an assert that it should never be null?

sure

**Comment by zanderso** [comment]
From PR review triage: @GuaiYiHu are you interested in continuing with this PR? If so, it looks like the next step is to follow-up on the feedback from @chunhtai.

**Comment by GuaiYiHu** [comment]
> From PR review triage: @GuaiYiHu are you interested in continuing with this PR? If so, it looks like the next step is to follow-up on the feedback from @chunhtai.

busy these days, will do this on my holiday from tomorrow.😃

**Comment by GuaiYiHu** [comment]
> From PR review triage: @GuaiYiHu are you interested in continuing with this PR? If so, it looks like the next step is to follow-up on the feedback from @chunhtai.

@chunhtai @zanderso all done, please review.

File: shell/platform/android/io/flutter/view/AccessibilityBridge.java [hunk 2]
```diff
@@ -2695,6 +2695,9 @@ private void updateRecursively(
         if (globalTransform == null) {
           globalTransform = new float[16];
         }
+        if (transform == null) {
```
Comment (chunhtai) [inline shell/platform/android/io/flutter/view/AccessibilityBridge.java:2698]: I imagine you still need another condition to initialize it  so it won't crash in release code.

I did try to compile a custom engine in release mode, but it looks like the assertion will still crash the app. @GaryQian Do you know if the assertionError will be ignored in release mode?

**Review by chunhtai — COMMENTED** [review]
Add a comment

File: shell/platform/android/io/flutter/view/AccessibilityBridge.java (hunk 2 above)
Comment (GuaiYiHu) [reply shell/platform/android/io/flutter/view/AccessibilityBridge.java:2698]: You meant the exception will be thrown only on debug mode, right? Shall BuildConfig.DEBUG work?

Comment (chunhtai) [reply shell/platform/android/io/flutter/view/AccessibilityBridge.java:2698]: yeah I think BuildConfig.DEBUG is probably the best way to do this

Comment (GuaiYiHu) [reply shell/platform/android/io/flutter/view/AccessibilityBridge.java:2698]: will reedit it.

**Comment by GuaiYiHu** [comment]
@chunhtai Pushed again, all checks have passed, please review, thanks a lot.

File: shell/platform/android/io/flutter/view/AccessibilityBridge.java [hunk 3]
```diff
@@ -2695,6 +2695,13 @@ private void updateRecursively(
...
+          if (BuildConfig.DEBUG) {
+            throw new AssertionError("transform has not been initialized");
+          } else {
```
Comment (chunhtai) [inline shell/platform/android/io/flutter/view/AccessibilityBridge.java:2701]: this else case is redundant 
```java
if (BuildConfig.DEBUG) {
  throw new AssertionError("transform has not been initialized");
}
transform = new float[16];

```

**Review by chunhtai — APPROVED** [review]
LGTM, I don't think there is a good way to test this change. This is adding an instrument to catch something should happen in the first place.

**Comment by chunhtai** [comment]
Can you add a test to test if it throw assertionError when a node has uninitialized transform, and mock the BuildConfig to test it doesn't crash?

**Comment by GuaiYiHu** [comment]
> Can you add a test to test if it throw assertionError when a node has uninitialized transform, and mock the BuildConfig to test it doesn't crash?

I'm afraid I cannot reproduce this crash, it appears just occasionally.

**Comment by chunhtai** [comment]
> > Can you add a test to test if it throw assertionError when a node has uninitialized transform, and mock the BuildConfig to test it doesn't crash?
> 
> I'm afraid I cannot reproduce this crash, it appears just occasionally.

Can you create test similar to ones in AccessibilityBridgeText that sends a TestSemanticsUpdate with a node that contains a child id but not part of the update, for example, you can duplicate TestSemanticsNode but in the `addToBuffer` method comment out the last part that add child to the update

**Comment by GaryQian** [comment]
@GuaiYiHu Did you still plan on adding a test or finding a way to repro this bug?

**Comment by chinmaygarde** [comment]
From Triage: This looks like a benign change that seems to be hard to test and follows patterns consistent with the rest of the code. It's low risk, fixes crashes found in the wild and has an LGTM. Let's land this and file followup issues as needed.

**Review by chinmaygarde — APPROVED** [review]

**Pull request merged**